Session sequence number

adamhass opened the pull request:
Please make sure these boxes are checked, before submitting a new PR.

- [x] You ran `rustfmt` on the code base before submitting (on latest nightly with rustfmt support)
- [x] You reference which issue is being closed in the PR text (if applicable)

## Issues

Closes #84

## Other Changes

* Added `SessionId` structure ( a wrapper for `Uuid`'s)
* Added `session: Option<SessionId>` field to `NetMessage`
* Added `SessionId` to three `NetworkStatus` messages: `ConnectionEstablished`, `ConnectionLost`, and `ConnectionClosed`. 

SessionId's are described in the doc of the `NetMessage` struct:

````
    /// Each physical end-to-end TCP connection between two systems is assigned a unique identifier.
    /// When a connection is lost/closed and then re-established, the connection will have a
    /// new unique identifier. The unique identifier is negotiated and is the same on both
    /// ends of the connection.
    ///
    /// The field is only set by the Network-layer in incoming TCP messages.
    /// All other messages will have the field set to `None`.
    ///
    /// For any sequence of two messages received from a remote actor over TCP:
    ///     If the session of the messages differs, an intermediate message *may* have been lost.
    ///     Conversely, if the session does not differ no intermediate message was lost.
````

Bathtor commented:
Ok, I probably won't have time for a proper review, so I'll leave that to Harald, but just a question basically from the PR description along:

1) Are we sure `u8` is going be sufficient as a size for the session id? 256 sessions and the wrap-around I guess?
2) What are your concerns about using a larger type like `u32`? Is it storage size for our open connections data structure? Because we can't have more open connections at any given time than there are possible ports (65k) which means the difference between storing `u8` and `u32` is only 196k which I would vaguely label as irrelevant. That isn't even taking possible alignment issues into account.

adamhass commented:
> Ok, I probably won't have time for a proper review, so I'll leave that to Harald, but just a question basically from the PR description alone:
> 
> 1. Are we sure `u8` is going be sufficient as a size for the session id? 256 sessions and then wrap-around I guess?
> 2. What are your concerns about using a larger type like `u32`? Is it storage size for our open connections data structure? Because we can't have more open connections at any given time than there are possible ports (65k) which means the difference between storing `u8` and `u32` is only 196k which I would vaguely label as irrelevant. That isn't even taking possible alignment issues into account.

My concern about storage size is because it's included in every NetMessage. It's likely irrelevant. My first idea was to use the already existing UUID's associated with channels, but I figured that they're A: not sorted and B: 128bit overhead with each NetMessage felt like pushing it. I'm absolutely fine with changing it to a u32 though.

haraldng commented:
Adam and I discussed this today and we agree that the channel id i.e. UUID is enough.

We were also discussing if there might be a somewhat unneccessary overhead of including a UUID in every NetMessage, and it therefore might make sense to have this as a feature flag instead. We're not sure if that 128 bit overhead matters for local performance... What do you think @Bathtor?

Bathtor commented:
> Adam and I discussed this today and we agree that the channel id i.e. UUID is enough.
> 
> We were also discussing if there might be a somewhat unneccessary overhead of including a UUID in every NetMessage, and it therefore might make sense to have this as a feature flag instead. We're not sure if that 128 bit overhead matters for local performance... What do you think @Bathtor?

I honestly don't think it'll matter. Let's just use it for now, and if we ever figure out that it is indeed an issue we can always go back and re-evaluate our decision.

adamhass commented:
Thank you both for the feedback. I've reworked the PR now and I changed the top comment of this PR to reflect what it now does.

The NetMessage struct has a `Option<SessionId>` field. If the `NetMessage` is received on a TCP Channel the Option will be set to the `SessionId` of the TCP Channel it was received on. Both sides of the channel have the same SessionId, which may be useful in debugging and tracing distributed deployments.

NetMessages that are routed locally, or received over UDP will have the option set to `None`.

haraldng commented on core/src/messaging/net_message.rs:
```diff
@@ -132,6 +149,16 @@ impl NetMessage {
+    }
+
+    /// Sets the SessionId of the `NetMessage`
+    pub fn set_session(&mut self, session: SessionId) -> () {
```
I think this function should be hidden from the user. Perhaps use pub(crate) or whatever that is needed to be used internally.

haraldng replied:
I think this function should be hidden from the user. Perhaps use pub(crate) or whatever that is needed to be used internally.

haraldng reviewed (commented):
Looks good. I have only one small comment regarding visibility.

Bathtor commented on core/src/messaging/net_message.rs:
```diff
@@ -132,6 +149,16 @@ impl NetMessage {
+        self.session
+    }
+
+    /// Sets the SessionId of the `NetMessage`
+    pub(crate) fn set_session(&mut self, session: SessionId) -> () {
+        self.session = Some(session);
+    }
```
I actually don't necessary agree with Harald here. Remember that people could be writing 3rd-party network implementations? This makes it impossible for those to support session ids.

Bathtor commented on core/src/net/mod.rs:
```diff
@@ -54,6 +55,34 @@ impl From<Transport> for Protocol {
+/// Session identifier, part of the `NetMessage` struct. Managed by Kompact internally, may be read
+/// by users to detect session loss, indicated by different SessionId's of NetMessages.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct SessionId {
+    session: Uuid,
+}
```
I'd prefer a simple new-type without the named field.

Bathtor commented on core/src/net/mod.rs:
```diff
@@ -54,6 +55,34 @@ impl From<Transport> for Protocol {
+
+impl SessionId {
+    /// Generates a new SessionId
+    pub(crate) fn new() -> SessionId {
+        SessionId {
+            session: Uuid::new_v4(),
+        }
+    }
```
Again this makes it impossible to use in 3rd party impls.

Bathtor replied:
Also shall we call this `new_random` or `new_unique` to make it clear at use-site that calling this twice will give different results?

Bathtor commented on core/tests/dispatch_integration_tests.rs:
```diff
@@ -1484,7 +1499,7 @@ fn network_status_port_block_unblock_ip() {
         .wait_timeout(PINGPONG_TIMEOUT)
         .expect("Time out waiting for ping pong to complete");
 
-    status_counter.on_definition(|sc| assert_eq!(sc.blocked_ip.contains(&pinger_ip), false));
```
How did I overlook this awkwardness before? :D

Bathtor reviewed (changes requested):
Found some time to do a review afterall while waiting for other stuff to finish.

In addition to the inline comments: Can we make clippy changes a separate PR next time? It really pollutes this one, and costs extra reviewer time. (I guess I'm getting more picky on this now that I'm a) doing a lot more reviews a week and b) not having as much time for Kompact^^)

adamhass commented on core/src/dispatch/lookup/path_trie.rs:
```diff
@@ -340,9 +340,8 @@ mod tests {
 
-        let values: Vec<usize> = trie.values().copied().collect();
         for i in 0..=5 {
-            assert!(values.contains(&i));
+            assert!(trie.values().copied().any(|v| { v == i }));
         }
```
It's a mess I know. Clippy doesn't like collect, wants me to use the iterator but I can't search through an iterator multiple times. I can remove the `copied()` call though.

Bathtor commented on core/src/net/mod.rs:
```diff
@@ -54,6 +55,28 @@ impl From<Transport> for Protocol {
 
+/// Session identifier, part of the `NetMessage` struct. Managed by Kompact internally, may be read
+/// by users to detect session loss, indicated by different SessionId's of NetMessages.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct SessionId(Uuid);
```
Oh, I overlooked this before: Maybe we should **not** derive `PartialOrd` and `Ord` here, since we are specifically not promising session id ordering as discussed before. I feel like giving people the ability to do `id_a < id_b` would suggest that this a sensible operation, when it really is not.

adamhass replied:
Good catch, fully agree, leaving it there after changing to uuid was not intentional

Bathtor commented on core/src/net/mod.rs:
```diff
@@ -54,6 +55,28 @@ impl From<Transport> for Protocol {
+        SessionId(Uuid::new_v4())
+    }
+
+    /// For Serialisation of SessionId's
+    pub fn as_u128(&self) -> u128 {
+        self.0.as_u128()
+    }
+
+    /// For Deserialisation of SessionId's
+    pub fn from_u128(v: u128) -> SessionId {
+        SessionId(Uuid::from_u128(v))
+    }
```
Can we give these a big warning in the docs? Stating that they are only exposed for the purpose of writing custom serialisers, but no other code should rely on this as we are not promising the internal representation of session ids to stay compatible with this. Otherwise our previous discussion of "meh, we can change them later if we feel the need" is not really going to be true^^

Bathtor reviewed (approved):
Just two more tiny comments and then it's ready to merge.
